### PR TITLE
[Demangler] Don't truncate demangling of an associated type descriptor

### PR DESCRIPTION
**Pull request by theblixguy**
When demangling an associated type descriptor name like `$s5Assoc3Kit5ProtoPTl`, the associated type's name is printed, but information about the protocol it belongs to is not, even though it's available in the mangled name. When we have information about the protocol that the type belongs to, print that as well.

```
$ swift demangle -expand s5Assoc3Kit5ProtoPTl
Demangling for $s5Assoc3Kit5ProtoPTl
kind=Global
  kind=AssociatedTypeDescriptor
    kind=DependentAssociatedTypeRef
      kind=Identifier, text="Assoc"
      kind=Type
        kind=Protocol
          kind=Module, text="Kit"
          kind=Identifier, text="Proto"
$s5Assoc3Kit5ProtoPTl ---> associated type descriptor for Assoc
```

Resolves [SR-11403](https://bugs.swift.org/browse/SR-11403).

### CODE REVIEW

**Comment by theblixguy** [comment]
@swift-ci please smoke test

File: lib/Demangling/NodePrinter.cpp [hunk 1]
```diff
@@ -1704,10 +1704,16 @@ NodePointer NodePrinter::print(NodePointer Node, bool asPrefixContext) {
     Printer << ": ";
     print(Node->getChild(2));
     return nullptr;
-  case Node::Kind::AssociatedTypeDescriptor:
+  case Node::Kind::AssociatedTypeDescriptor: {
     Printer << "associated type descriptor for ";
-    print(Node->getChild(0));
+    auto assoc = Node->getChild(0);
+    if (auto parent = getFirstChildOfKind(assoc, Node::Kind::Type)) {
```
Comment (brentdax) [inline lib/Demangling/NodePrinter.cpp]: It’s kind of strange that you need to reach into a grandchild node here. Would it make sense to modify the way the child node is printed instead?

**Review by beccadax — CHANGES_REQUESTED** [review]

File: lib/Demangling/NodePrinter.cpp (hunk 1 above)
Comment (theblixguy) [reply lib/Demangling/NodePrinter.cpp]: Sure, that's actually better. I have made the change in feef2da.

**Comment by brentdax** [comment]
@swift-ci please smoke test

**Review by beccadax — APPROVED** [review]
Nice and clean now!

**Comment by theblixguy** [comment]
Hmm, it seems like we may not have the extra node in some cases, so we have to guard against that. Also, it seems like some of the tests failed because the demangled string was different due to this change.

Let me update the tests.

**Comment by theblixguy** [comment]
@swift-ci please smoke test macOS platform

**Comment by theblixguy** [comment]
@swift-ci please smoke test macOS platform

**Comment by theblixguy** [comment]
I wish FileCheck would dump all the failures at once instead of one at a time :/ 

I’ve been having some issues with running tests locally (kernel panics on Catalina beta), but let me try to resolve them locally and then push my changes.

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Comment by brentdax** [comment]
You know, looking at the test cases you changed, I'm not sure this is the best way to print these things. For example:

```
_TtuRxs8Runciblexs8FungiblerFxwxPS_5Mince ---> <A where A: Swift.Runcible, A: Swift.Fungible>(A) -> A.Swift.Runcible.Mince
```

Might be better rendered as something like `A.Mince (from Swift.Runcible)`. It's a shame we don't already have a syntax for this that we could crib from...

**Comment by theblixguy** [comment]
I think we do something similar for extensions, for example: “A.Swift._IndexableBase.Index in (extension in a)”.

**Comment by brentdax** [comment]
It looks like we *now* do that after this PR, but we didn't before it.

**Comment by theblixguy** [comment]
Yeah, I meant the “in (extension in a)” bit.

**Comment by brentdax** [comment]
Right. I meant a syntax in Swift itself—for instance, if the language allowed you to say something like `Runcible::Mince` to distinguish it from a `Mince` associated type you got from some other protocol, that might give us an idea of how to spell this thing in a demangling.

**Pull request merged**